File: ios/DailyBucketUtils.swift
import Foundation

struct DailyBucketRange {
    let startTimeMs: Double
    let endTimeMs: Double
}

func clampDailyBucketRange(
    bucketStartTimeMs: Double,
    bucketEndTimeMs: Double,
    queryStartTimeMs: Double,
    queryEndTimeMs: Double
) -> DailyBucketRange {
    let startTimeMs = min(max(bucketStartTimeMs, queryStartTimeMs), queryEndTimeMs)

    return DailyBucketRange(
        startTimeMs: startTimeMs,
        // Guard against inverted ranges when a bucket falls outside the query
        // range. Mirrors clampDailyBucketRange on Android.
        endTimeMs: max(min(bucketEndTimeMs, queryEndTimeMs), startTimeMs)
    )
}

File: android/src/main/java/com/nitrohealth/DailyBucketUtils.kt
package com.nitrohealth

internal data class DailyBucketRange(
    val startTimeMs: Double,
    val endTimeMs: Double
)

internal fun clampDailyBucketRange(
    bucketStartTimeMs: Double,
    bucketEndTimeMs: Double,
    queryStartTimeMs: Double,
    queryEndTimeMs: Double
): DailyBucketRange {
    val startTimeMs = bucketStartTimeMs.coerceIn(queryStartTimeMs, queryEndTimeMs)

    return DailyBucketRange(
        startTimeMs = startTimeMs,
        // Guard against inverted ranges when a bucket boundary round-trips across a
        // DST transition and lands outside the query range.
        endTimeMs = bucketEndTimeMs.coerceAtMost(queryEndTimeMs).coerceAtLeast(startTimeMs)
    )
}
